Host panic - nomad 0.5.2 - client.runAllocs

Host panic, and nomad was unable to restart afterwords.

```
.....
    2017/01/23 17:30:12.568052 [DEBUG] client: allocs: (added 4) (removed 0) (updated 22) (ignore 18)
    2017/01/23 17:30:12.568193 [DEBUG] client: added alloc "c54717a6-7b63-428f-d569-a8a3553907d2" to blocked queue for previous allocation "c294bf88-8afa-5e6a-e123-14a671d77fa5"
    2017/01/23 17:30:12.584264 [DEBUG] client: allocs: (added 4) (removed 0) (updated 22) (ignore 18)
panic: close of closed channel

goroutine 131 [running]:
panic(0xfc0740, 0xc45c60b5f0)
        /opt/go/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/nomad/client.(*Client).runAllocs(0xc420080dc0, 0xc48f1ca880)
        /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:1369 +0xde0
github.com/hashicorp/nomad/client.(*Client).run(0xc420080dc0)
        /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:927 +0x119
created by github.com/hashicorp/nomad/client.NewClient
        /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:266 +0xc78
    Loaded configuration from /etc/nomad/config.hcl
==> Starting Nomad agent...
==> Error starting agent: client setup failed: failed to restore state: 1 error(s) occurred:

* 1 error(s) occurred:

* alloc_runner snapshot includes a nil context
    Loaded configuration from /etc/nomad/config.hcl

```
### Nomad version
Output from `nomad version`
Nomad v0.5.2

### Operating system and Environment details
Linux ip-10-XXX-X-XXX 4.4.0-59-generic #80-Ubuntu SMP Fri Jan 6 17:47:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### Issue
Runtime panic. Developer was using the logs API endpoint heavily during this time.

### Reproduction steps
Not sure -it was otherwise normal usage. Heavy use of logs, however.

### Nomad Server logs (if appropriate)

### Nomad Client logs (if appropriate)
[client.zip](https://github.com/hashicorp/nomad/files/724492/client.zip)

### Job file (if appropriate)

--- a/client/client.go
+++ b/client/client.go
@@ -150,12 +150,38 @@ type Client struct {
 	vaultClient vaultclient.VaultClient
 
 	// migratingAllocs is the set of allocs whose data migration is in flight
-	migratingAllocs     map[string]chan struct{}
+	migratingAllocs     map[string]*migrateAllocCtrl
 	migratingAllocsLock sync.Mutex
 
 	// garbageCollector is used to garbage collect terminal allocations present
 	// in the node automatically
 	garbageCollector *AllocGarbageCollector
+}
+
+// migrateAllocCtrl indicates whether migration is complete
+type migrateAllocCtrl struct {
+	ch     chan struct{}
+	closed bool
+	chLock sync.Mutex
+}
+
+func newMigrateAllocCtrl() *migrateAllocCtrl {
+	return &migrateAllocCtrl{
+		ch: make(chan struct{}),
+	}
+}
+
+func (m *migrateAllocCtrl) closeCh() {
+	m.chLock.Lock()
+	defer m.chLock.Unlock()
+
+	if m.closed {
+		return
+	}
+
+	// If channel is not closed then close it
+	m.closed = true
+	close(m.ch)
 }
 
 var (
@@ -188,7 +214,7 @@ func NewClient(cfg *config.Config, consulSyncer *consul.Syncer, logger *log.Logg
 		blockedAllocations:  make(map[string]*structs.Allocation),
 		allocUpdates:        make(chan *structs.Allocation, 64),
 		shutdownCh:          make(chan struct{}),
-		migratingAllocs:     make(map[string]chan struct{}),
+		migratingAllocs:     make(map[string]*migrateAllocCtrl),
 		servers:             newServerList(),
 		triggerDiscoveryCh:  make(chan struct{}),
 		serversDiscoveredCh: make(chan struct{}),
@@ -1420,7 +1446,7 @@ func (c *Client) runAllocs(update *allocUpdates) {
 			// Stopping the migration if the allocation doesn't need any
 			// migration
 			if !update.updated.ShouldMigrate() {
-				close(ch)
+				ch.closeCh()
 			}
 		}
 	}
@@ -1455,7 +1481,7 @@ func (c *Client) runAllocs(update *allocUpdates) {
 				// prevents a race between a finishing blockForRemoteAlloc and
 				// another invocation of runAllocs
 				if _, ok := c.getAllocRunners()[add.PreviousAllocation]; !ok {
-					c.migratingAllocs[add.ID] = make(chan struct{})
+					c.migratingAllocs[add.ID] = newMigrateAllocCtrl()
 					go c.blockForRemoteAlloc(add)
 				}
 			}
@@ -1533,7 +1559,7 @@ ADDALLOC:
 
 // waitForAllocTerminal waits for an allocation with the given alloc id to
 // transition to terminal state and blocks the caller until then.
-func (c *Client) waitForAllocTerminal(allocID string, stopCh chan struct{}) (*structs.Allocation, error) {
+func (c *Client) waitForAllocTerminal(allocID string, stopCh *migrateAllocCtrl) (*structs.Allocation, error) {
 	req := structs.AllocSpecificRequest{
 		AllocID: allocID,
 		QueryOptions: structs.QueryOptions{
@@ -1551,7 +1577,7 @@ func (c *Client) waitForAllocTerminal(allocID string, stopCh chan struct{}) (*st
 			select {
 			case <-time.After(retry):
 				continue
-			case <-stopCh:
+			case <-stopCh.ch:
 				return nil, fmt.Errorf("giving up waiting on alloc %v since migration is not needed", allocID)
 			case <-c.shutdownCh:
 				return nil, fmt.Errorf("aborting because client is shutting down")
@@ -1665,7 +1691,7 @@ func (c *Client) unarchiveAllocDir(resp io.ReadCloser, allocID string, pathToAll
 	for {
 		// See if the alloc still needs migration
 		select {
-		case <-stopMigrating:
+		case <-stopMigrating.ch:
 			os.RemoveAll(pathToAllocDir)
 			c.logger.Printf("[INFO] client: stopping migration of allocdir for alloc: %v", allocID)
 			return nil

